logendpoint.cpp

```diff
@@ -0,0 +1,213 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "logendpoint.h"
+
+#include <dirent.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "log.h"
+#include "mainloop.h"
+#include "util.h"
+
+#define MAX_LOG_FILE_PREFIX 999
+
+static bool _logging_start_timeout_cb(void *data)
+{
+    LogEndpoint *log = static_cast<LogEndpoint *>(data);
+    return log->_start_timeout();
+}
+
+static bool _alive_timeout_cb(void *data)
+{
+    LogEndpoint *log = static_cast<LogEndpoint *>(data);
+    return log->_alive_timeout();
+}
+
+void LogEndpoint::_send_msg(const mavlink_message_t *msg, int target_sysid)
+{
+    uint8_t data[MAVLINK_MAX_PACKET_LEN];
+    struct buffer buffer {
+        0, data
+    };
+
+    buffer.len = mavlink_msg_to_send_buffer(data, msg);
+    _mainloop->route_msg(&buffer, target_sysid, msg->sysid);
+
+    _stat.read.total++;
+    _stat.read.handled++;
+    _stat.read.handled_bytes += msg->len;
+    if (msg->magic == MAVLINK_STX)
+        _stat.read.handled_bytes += sizeof(struct mavlink_router_mavlink2_header);
+    else
+        _stat.read.handled_bytes += sizeof(struct mavlink_router_mavlink1_header);
+}
+
+int LogEndpoint::_get_prefix(const char *extension, const char *dirname, uint16_t *prefix)
+{
+    struct dirent *ent;
+    DIR *dir;
+    uint16_t u;
+
+    dir = opendir(dirname);
+    if (!dir) {
+        return -errno;
+    }
+
+    *prefix = 0;
+
+    while ((ent = readdir(dir)) != nullptr) {
+        if (sscanf(ent->d_name, "%hu-", &u) == 1) {
+            if (u >= *prefix && u < MAX_LOG_FILE_PREFIX) {
+                *prefix = ++u;
+            }
+        }
+    }
+
+    closedir(dir);
+
+    return 0;
+}
+
+int LogEndpoint::_get_file(const char *extension, char **filename_result)
+{
+    char filename[PATH_MAX] = {};
+
+    time_t t = time(NULL);
+    struct tm *timeinfo = localtime(&t);
+    int r;
+    uint16_t i;
+
+    if (_get_prefix(extension, _logs_dir, &i) < 0) {
+        log_error_errno(errno, "Error formatting Log file name: (%m)");
+    }
+
+    for (; i <= MAX_LOG_FILE_PREFIX; i++) {
+        r = snprintf(filename, sizeof(filename), "%s/%03u-%i-%02i-%02i_%02i-%02i-%02i.%s",
+                     _logs_dir, i, timeinfo->tm_year + 1900, timeinfo->tm_mon + 1,
+                     timeinfo->tm_mday, timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec,
+                     extension);
+
+        if (r < 1 || (size_t)r >= sizeof(filename)) {
+            log_error_errno(errno, "Error formatting Log file name: (%m)");
+            return -EINVAL;
+        }
+
+        r = open(filename, O_WRONLY | O_CLOEXEC | O_CREAT | O_NONBLOCK | O_TRUNC,
+                 S_IRUSR | S_IROTH | S_IRGRP);
+        if (r >= 0) {
```

Comment:
return early on error, not on success.... the only exception is if it makes way simpler to handle the success case before, which is not in this case.

btw the for above should be a fixed number of retries, not the number available in the prefix range